compiler/test/dotty/tools/vulpix/ParallelTesting.scala

```diff
@@ -522,49 +541,64 @@ trait ParallelTesting { self =>
-              echoBuildInstructions(reporter, testSource, reporter.errorCount, reporter.warningCount)
+            if (reporter.compilerCrashed || reporter.errorCount > 0) {
+              logReporterContents(reporter)
+              logBuildInstructions(reporter, testSource, reporter.errorCount, reporter.warningCount)
+            }
 
-            registerCompilation(reporter.errorCount)
-            (reporter.errorCount, reporter.warningCount, checkFile.isDefined, () => verifyOutput(checkFile.get, outDir, testSource, reporter.warningCount))
+            (reporter.compilerCrashed, reporter.errorCount, reporter.warningCount, checkFile.isDefined, () => verifyOutput(checkFile.get, outDir, testSource, reporter.warningCount))
           }
 
           case testSource @ SeparateCompilationSource(_, dir, flags, outDir) => {
             val checkFile = new JFile(dir.getAbsolutePath.reverse.dropWhile(_ == '/').reverse + ".check")
+            val reporters = testSource.compilationGroups.map(compile(_, flags, false, outDir))
+            val compilerCrashed = reporters.exists(_.compilerCrashed)
             val (errorCount, warningCount) =
-              testSource
-                .compilationGroups
-                .map(compile(_, flags, false, outDir))
-                .foldLeft((0,0)) { case ((errors, warnings), reporter) =>
-                  if (reporter.errorCount > 0)
-                    echoBuildInstructions(reporter, testSource, reporter.errorCount, reporter.warningCount)
+              reporters.foldLeft((0,0)) { case ((errors, warnings), reporter) =>
+                if (reporter.errorCount > 0)
+                  logBuildInstructions(reporter, testSource, reporter.errorCount, reporter.warningCount)
 
-                  (errors + reporter.errorCount, warnings + reporter.warningCount)
-                }
+                (errors + reporter.errorCount, warnings + reporter.warningCount)
+              }
 
-            if (errorCount > 0) fail()
+            if (errorCount > 0) {
+              reporters.foreach(logReporterContents)
+              logBuildInstructions(reporters.head, testSource, errorCount, warningCount)
+            }
 
-            registerCompilation(errorCount)
-            (errorCount, warningCount, checkFile.exists, () => verifyOutput(checkFile, outDir, testSource, warningCount))
+            (compilerCrashed, errorCount, warningCount, checkFile.exists, () => verifyOutput(checkFile, outDir, testSource, warningCount))
           }
         }
 
-        if (errorCount == 0 && hasCheckFile) verifier()
-        else if (errorCount == 0) runMain(testSource.outDir, testSource)
-        else if (errorCount > 0) {
-          echo(s"\nCompilation failed for: '$testSource'")
+        if (!compilerCrashed && errorCount == 0 && hasCheckFile) verifier()
+        else if (!compilerCrashed && errorCount == 0) {
+          if (Properties.testsNoRun) addNoRunWarning()
+          else runMain(testSource.classPath) match {
```

Comment:
I don't like the repetition with `verifyOutput` above here, would be nicer if we pretended that all tests had a check file (no check file would be equivalent to empty check file).